File: app/src/main/java/com/hencoder/hencoderpracticedraw1/practice/Practice3DrawRectView.java
package com.hencoder.hencoderpracticedraw1.practice;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.RectF;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;

import com.hencoder.hencoderpracticedraw1.Utils;

public class Practice3DrawRectView extends View {
    private static final String TAG = "Practice3DrawRectView";
    RectF mRectF = new RectF();
    Paint mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);

    public Practice3DrawRectView(Context context) {
        super(context);
    }

    public Practice3DrawRectView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

    public Practice3DrawRectView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    {
        int screenWidth = Utils.getScreenWidth();
        Log.d(TAG, "instance initializer: screenWidth=" + screenWidth);
        float side = Utils.dp2px(70);
        mRectF.set((screenWidth >> 1) - side, 100, (screenWidth >> 1) + side, 100 + 2 * side);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

//        练习内容：使用 canvas.drawRect() 方法画矩形
        canvas.drawRect(mRectF, mPaint);
    }
}

File: app/src/main/java/com/hencoder/hencoderpracticedraw1/practice/Practice3DrawRectView.kt
package com.hencoder.hencoderpracticedraw1.practice

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Rect
import android.util.AttributeSet
import android.view.View

/**
 * Created by xh on 2017/10/16.
 */
class Practice3DrawRectView : View {

    private var mPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    private var mRect = Rect()

    private val mLength = 400

    private var mLeft = 0

    private var mTop = 0

    constructor(context : Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        super.onSizeChanged(w, h, oldw, oldh)

        mLeft = (w - mLength) / 2
        mTop = (h - mLength) / 2

        mRect.set(mLeft, mTop, mLeft + mLength, mTop + mLength)

    }

    override fun draw(canvas: Canvas?) {
        super.draw(canvas)

        canvas!!.drawRect(mRect, mPaint)

    }

}